### ISSUE TITLE
BUG: Unwanted file deletion when mass updadting files while Zettlr is open

### ISSUE THREAD

**Issue by Jaifroid** [Zettlr/Zettlr#51]
## Description
When Zettlr is open and showing a document in a project directory, and if another process mass-updates files (e.g. mass-reverting uncommitted changes in git, mass-updating files using find + sed in a bash shell, etc.), Zettlr seems to get overwhelmed with the number of changes, and deletes some of the files. Fortunately I have not lost files, because I have git running on my important markdown folder.

## Reproducing
Method 1: Have a git-controlled directory with a Zettlr project containing a large number of files in directories and subdirectories (I don't know how many required to reproduce, but I have more than 20). One of the files should be open in Zettlr. Each file should have a minor change you don't care about. Use git / GitHub desktop, to revert all changes at once. Zettlr asks if you want to reload the currently open file. Its screen fills with "File changed by external process" (or similar message). Shortly after, check in Git, and some of the files (not all) will have been deleted subsequent to the revert.

Method 2: Have a set of 20 or so utf8 markdown files with a BOM in a project directory that is open in Zettlr (with one of the files open in editor). Using a bash shell (I'm using WSL), run 
`find -iname \*.md -type f -exec sed -i '1s/^\(\xef\xbb\xbf\)\?//' {} +`
(This removes the BOM, if present, on all files.) Zettlr detects the change and asks to reload current file, while its display fills with messages about changed files. Check in Git. Apart from the changes to each file, you will see that some of the files are also deleted.

The deletions do not occur if Zettlr is closed.

## Expected behaviour
Reload the currently open file, and not touch the others.

## Platform
 - OS: Windows 10
 - OS Version: 1803
 - Zettlr Version: 0.20

## Additional information
I only tested under Windows. I reproduced it when mass-updating in three scenarios: mass-updating with GitHub Desktop; mass-updating using the WSL bash shell; mass-updating files using a PowerShell script. In each case, the deletions occur only while Zettlr is open.

## Workaround
Remember to close Zettlr before doing mass operations on files!

**Comment by nathanlesage** [comment]
Uh oh, this should not be happening. I'm currently unable to try this at home because I have some other deadlines approaching (and leaving Zettlr closed alleviates the problem for now), but I have a theory why this happens.

The watchdog code that monitors the file system for changes is about 6 months old, and I wrote it at a time where I didn't know all the concepts of coding in node. This resulted in a crude structure, in which all changes to the file system are buffered in an array and every 5 seconds the app goes through that array, handling all changes at once. This was back then the only way I saw fit for dealing with remote changes.

But one problem remains: If the current file in the editor is changed by this, it will ask you to replace it, and I think there's the problem happening: As soon as Zettlr recognises a remote change in the current file, it will prompt you with a modal that is effectively **blocking the main process until a button is clicked**. And as all file modifications are processed in the same tick of the application, this might be the cause (i.e. some file descriptors not being released due to the block).

This means that maybe Git (or a shell script or whatever) will try to modify some files that are already opened by another process, which they might not expect, resulting in strange behaviour such as Git removing files. The files are definitely not removed by Zettlr, because there are enough safeguards against this, but I can definitely think of Zettlr forcing Git to do strange things when there are open file descriptors.

TL;DR: Would you be able to double-check the problem by _not_ having open a file affected by these mass edits? If this runs smoothly, then this would mean that the block is the problem.

Cheers!

P.S.: I know all this because I'm currently elaborating on ways of how to refactor the watchdog code so that all of these problems are (hopefully) solved in one of the next versions!

**Comment by nathanlesage** [comment]
Okay, I just made the modal non-blocking that asks you to replace your files. I've then tried it with my own Zettelkasten (89 files). Nothing missing after resetting the git repository, so making the modal non-blocking seemed to have done the job!

If you want to, you can for sure test the current dev-branch, to see if you can confirm my observations.

Please tell me whether or not it worked, so that I can close the issue!

**Comment by nathanlesage** [comment]
Nope, that didn't help, just realised 5 files were missing here as well … maybe the whole re-reading thing is just too slow after all

**Comment by nathanlesage** [comment]
Another update: I just found the five missing files in the trash :D Could you please check if the files are there on your computer? If so, we would've circled in on the problem much more, because there is only one place inside the code where Zettlr actively moves the files to the trash.

**Comment by nathanlesage** [comment]
Fixed the bug that files will push themselves into the trash while they are simply handling an event. Now I only need a solution for directories as well!

**Comment by Jaifroid** [comment]
Thanks for working on this! How do I test the dev branch? Do I need to build it myself with node etc? Sorry, I've just been using Zettlr with the released version (self-updating).

I never really checked in Recycle Bin, because I was using Git control, so simply reverted all changes when I realized files were missing.

I *may* get some time this weekend, but it's looking quite heavy, and I may not be able to help with testing till next week. But rest assured I'm interested in the outcome, as I've been enjoying using Zettlr for a real-life project.

**Comment by nathanlesage** [comment]
@Jaifroid That's very nice to hear :)

To test you would need to clone the repository, checkout to the develop branch and then you'd need yarn or npm (which should automatically install node), then `yarn/npm install` in the directory and then simply `yarn/npm start` the application. No need to build for testing!

But I think this might not be necessary for this bug (albeit I would really appreciate having someone testing on Windows!) — as I told you there are only two lines of code in the whole of Zettlr code that actually removes files and dirs. Do you remember how I told you there were enough safeguards to ensure this wouldn't be called inadvertedly? Guess what: yep, I overlooked one of these safeguards. I only need to add an additional check for preventing directories deleting themselves inadvertedly, and then it should be safe again to perform mass reverts without closing Zettlr!

Have a productive (and also, hopefully, a relaxing) weekend!

**Comment by nathanlesage** [comment]
So, implemented the second safeguard. Now it should be safe to mass-rebase without having Zettlr closed! :)

**Comment by Jaifroid** [comment]
Great, thanks! I've managed to npm install the developer branch and run with electron.exe. So I tested mass-updating 29 markdown files in a Zettlr Project directory by removing the BOM on all files with a PowerShell script while one of the files was open/displayed in Zettlr. No files were deleted, all were correctly updated, and the file open in Zettlr asked me to reload with a dialog box.

I then reversed the operation, adding the BOM back to all markdown files in the project. Same result, none were deleted.

I then removed the BOM again, and tried mass-reverting the changes in Git. No files were deleted, though there was a lot more activity in Zettlr, with all files appearing to disappear while Git deleted them, and then re-appearing again. There were various messages about files being removed. However, they all reappeared again afterwards, and none are missing. The difference here is simply due to the way Git reverts changes: it deletes files in the Working Tree and then restores them.

So, in summary, with these tests I can no longer reproduce the issue. It appears to be fixed! As it is a critical bug, it might be worth leaving the issue open for a few days for real-world testing, though that's up to you.

**Comment by nathanlesage** [comment]
That's great to hear!

Also I can now tell you what happened until the fix:

1. Git removed a file it was about to revert.
2. The watchdog immediately registered the deletion event.
3. Git created the reverted version of the file.
4. The watchdog immediately registered the add event.
5. Zettlr ran a `poll()`-run, and the first event it saw was the deletion event.
6. The `ZettlrFile` model executed the necessary code to deal with the deletion event of the file it was representing, calling `remove()` (back then without the newly added flag). This removed the file by moving it to the trash.
7. The next pass of the `for`-loop had the add event. It passed it to the `ZettlrDirectory` in which the add-event was registered.
8. The `ZettlrDirectory` model re-scanned its contents to reflect the addition event. But as in the meantime the newly added file has already been deleted in the previous pass of the `for`-loop, it did not add it, because the file was obviously not present anymore (it has been deleted in step 6).
9. The file was missing after the Git-revert finished.

What I changed was to add a flag to the `remove()` functions (of both directories and files, b/c both are reverted the same way in git) which is now implicitly set to `false`. If the flag is set to `false`, Zettlr will not attempt to delete the file on disk. The flag has to be set explicit to `true`, which is only done in two functions now — the functions that are being called if you consciously want to delete a file or directory.

So far for the explanation =D I'll use GitHubs automatic-close functionality to leave this issue open until the next version is released (b/c until then the bug persists and your workaround is needed).

Thanks for your testing and have a great weekend!
